### DISCUSSION TITLE
SQLalchemy can't communicate to MySQL server upon app start

### DISCUSSION THREAD

**Discussion by ralfeus** [sqlalchemy/sqlalchemy#11007] [Usage Questions] [answered]
I'm having troubles with SQLalchemy in Flask app first minute after the app start (or restart). 
It looks like logger exceptions of a sort:
> sqlalchemy.exc.OperationalError: (MySQLdb._exceptions.OperationalError) (2006, 'MySQL server has gone away') <---- most often one

> sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.

> sqlalchemy.exc.NoSuchColumnError: "Could not locate column in row for column 'users.id'"

> sqlalchemy.exc.NoSuchColumnError: "Could not locate column in row for column 'payments.id'"

> sqlalchemy.exc.OperationalError: (MySQLdb._exceptions.OperationalError) (2013, 'Lost connection to MySQL server during query')

Then everything gets back to normal. It's not a critical issue but annoying. 

Connection URI looks like this: `"mysql+mysqldb://user:password@localhost/mydb?unix_socket=/var/run/mysqld/mysqld.sock"`
I tried to run warming up queries upon application start:
```
    with app.app_context():
        for _ in range(5):
            try:
                db.session.execute(statement)
                logger.info("DB connection is successfully established")
                return
            except Exception as e:
                logger.warning(e)
                time.sleep(1)
    raise Exception("Couldn't establish a DB connection")
```

It passes through just fine but then I see same issues. 

It doesn't happen in development environment, only in production where Flask app runs on uwsgi server. Is there a way to fix it?

**Comment by CaselIT** [comment]
Hi,

Does the db runs in a container that's deployed with the application?
Asking to check if the application is maybe starting faster than the db.

Otherwise most of these error seem to be network related

**Comment by ralfeus** [comment]
Hi Federico,
the DB server runs on the same server and is not containerized. During the
application restart DB continues running. So the DB remains accessible
whole the time before, during and after app restart.

Regards
Michal

**Reply by CaselIT** [reply]
other than trying to use pre_ping I'm not sure what it would be causing it.
SQLalchemy is otherwise just using the driver you specify, so maybe try asking for guidance also on their repository.

@zzzeek can you think at something that would cause the db not to be available only on application startup time, not db startup?

**Comment by ralfeus** [comment, marked as answer]
In fact I've found an answer here: https://stackoverflow.com/a/59492567/12502959
The postfork handler had to be defined to close any open connections in forked process passed from original one

**Reply by CaselIT** [reply]
it's also documented at https://docs.sqlalchemy.org/en/20/core/pooling.html#pooling-multiprocessing

**Reply by ralfeus** [reply]
Indeed. The page I've found referred to one mentioned by you. And it was describing pretty much my use case:
 

> Depending on specifics of the driver and OS, the issues that arise here range from non-working connections to **socket connections that are used by multiple processes concurrently, leading to broken messaging** (the latter case is typically the most common).
